spec/fastsync/version2/scheduler.tla

```diff
@@ -0,0 +1,580 @@
+       /\ outEvent' = noEvent
+       /\ UNCHANGED scheduler
+     ELSE
+       /\ LET hp == highPeers(scheduler, minH) IN
+          IF hp = {} THEN
+            /\ outEvent' = noEvent
+            /\ UNCHANGED scheduler
+          ELSE \E bestPeerID \in hp:
+            /\ LET res == markPending(scheduler, bestPeerID, minH) IN
+               IF res.err /= noErr THEN
+                 outEvent' = [type |-> "scSchedulerFail", error|-> res.err]
+               ELSE
+                 outEvent' = [type |-> "scBlockRequest", peerID |-> bestPeerID, height |-> minH]
+               /\ scheduler' = res.val
+  /\ UNCHANGED scRunning
+
+handleBlockResponse ==
+  /\ inEvent.type = "bcBlockResponse"
+  /\ LET res == markReceived(scheduler, inEvent.peerID, inEvent.height) IN
+     IF res.err /= noErr THEN
+       LET res1 == removePeer(scheduler, inEvent.peerID) IN
+       /\ outEvent' = [type |-> "scPeerError", peerID |-> inEvent.peerID, reason |-> res.err]
+       /\ scheduler' = res1.val
+     ELSE
+       /\ outEvent' = [type |-> "scBlockReceived", peerID |-> inEvent.peerID, block |-> inEvent.block]
+       /\ scheduler' = res.val
+  /\ UNCHANGED scRunning
+
+handleNoBlockResponse ==
+  /\ inEvent.type = "bcNoBlockResponse"
+  /\ IF (scheduler.peers = {} \/ scheduler.peerStates[inEvent.peerID] = "peerStateRemoved") THEN
+        /\ outEvent' = noEvent
+        /\ UNCHANGED scheduler
+     ELSE
+       LET res == removePeer(scheduler, inEvent.peerID) IN
+        /\ outEvent' = [type |-> "scPeerError", peerID |-> inEvent.peerID, reason |-> "errPeerNoBlock"]
+        /\ scheduler' = res.val
+  /\ UNCHANGED scRunning
+
+handleBlockProcessed ==
+  /\ inEvent.type = "pcBlockProcessed"
+  /\ IF inEvent.height /= scheduler.height THEN
+       /\ outEvent' = [type |-> "scSchedulerFail", reason |-> "errProcessedBlockEv"]
+       /\ UNCHANGED scheduler
+     ELSE
+       LET res == markProcessed(scheduler, inEvent.height) IN
+       IF res.err /= noErr THEN
+         /\ outEvent' = [type |-> "scSchedulerFail", reason |-> res.err]
+         /\ UNCHANGED scheduler
+       ELSE
+         /\ scheduler' = res.val
+         /\ IF allBlocksProcessed(scheduler') THEN
+              outEvent' = [type |-> "scFinishedEv", reason |-> "finished"]
+            ELSE
+              outEvent' = noEvent
+  /\ UNCHANGED scRunning
+
+handleBlockProcessError ==
+  /\ inEvent.type = "pcBlockVerificationFailure"
+  /\ IF scheduler.peers = {} THEN
+        /\ outEvent' = noEvent
+        /\ UNCHANGED scheduler
+     ELSE
+        LET res1 == removePeer(scheduler, inEvent.firstPeerID) IN
+        LET res2 == removePeer(res1.val, inEvent.secondPeerID) IN
+          /\ IF allBlocksProcessed(res2.val) THEN
+               outEvent' = [type |-> "scFinishedEv", reason |-> "finished"]
+             ELSE
+               outEvent' = noEvent
+          /\ scheduler' = res2.val
+  /\ UNCHANGED scRunning
+
+onNoAdvanceExp ==
+    /\ inEvent.type = "tNoAdvanceExp"
```

Comment:
I think that this event does not have the same semantic as time based one (like done in the code). In the code, only if no blocks are processed for a period of time, i.e., height is not being updated during a period of time, then we terminate. Maybe right way to provide the same semantic is acting upon second such event, where we remember current height at the first occurrence of this event, and then on the second occurrence we check if height is still the same in which case we terminate.

Reply:
In the code we have an time ticker (10msec) and we check the time since the last block is executed. If > advanceTimeout then we terminate. Here we just simulate a time ticket satisfying this condition.  I could make the change but i'm not sure what kind of bug or different condition we would catch since the logic is so simple.